File: built-in-services/riscv/examples/dex/erc20.js
function assert(condition, msg) {
    if (!condition) {
        throw new Error(msg);
    }
}

function print(obj, name) {
    var obj_str = JSON.stringify(obj);
    PVM.debug(name || 'obj', ':', obj_str);
}

function init(name, symbol, supply) {
    if (!PVM.is_init()) {
        throw "init can only be invoked by deploy function";
    }
    PVM.set_storage('name', name);
    PVM.set_storage('symbol', symbol);
    PVM.set_storage('supply', supply.toString());
    // set caller balance to supply
    const caller = PVM.caller();
    _set_balance(caller, supply);
}

function total_supply() {
    return PVM.get_storage('supply');
}

function _balance_key(account) {
    return 'balance:' + account;
}

function _set_balance(account, amount) {
    const key = _balance_key(account);
    PVM.set_storage(key, amount.toString());
}

function _transfer(sender, recipient, amount) {
    if (amount <= 0) {
        throw 'amount must be positive';
    }
    const from_balance = parseInt(balance_of(sender));
    const to_balance = parseInt(balance_of(recipient));
    // TODO: change to some kind of safe math
    from_balance -= amount;
    assert(from_balance > 0, 'balance not enough');
    to_balance += amount;
    _set_balance(sender, from_balance);
    _set_balance(recipient, to_balance);
}

function transfer(recipient, amount) {
    const caller = PVM.caller();
    print({caller, recipient, amount}, 'transfer');
    _transfer(caller, recipient, amount);
}

function balance_of(account) {
    account = account || PVM.caller();
    const key = _balance_key(account);
    const ret = PVM.get_storage(key);
    return ret || '0';
}

function _approve(owner, spender, amount) {
    PVM.set_storage('allowances:' + owner + ':' + spender, amount.toString());
}

function approve(spender, amount) {
    _approve(PVM.caller(), spender, amount);
}

function allowances(owner, spender) {
    return PVM.get_storage('allowances:' + owner + ':' + spender) || '0';
}

function transfer_from(sender, recipient, amount) {
    const caller = PVM.caller();
    const before_allowance = parseInt(allowances(sender, caller));
    const after_allowance = before_allowance - amount;
    if (after_allowance < 0) {
        throw 'allowances not enough';
    }
    _transfer(sender, recipient, amount);
    _approve(sender, caller, after_allowance);
}

function _main(args) {
    if (args.method == 'init') {
        init(args.name, args.symbol, args.supply);
    } else if (args.method == 'total_supply') {
        return total_supply();
    } else if (args.method == 'balance_of') {
        return balance_of(args.account);
    } else if (args.method == 'transfer') {
        return transfer(args.recipient, args.amount);
    } else if (args.method == 'allowances') {
        return allowances(args.owner, args.spender);
    } else if (args.method == 'approve') {
        return approve(args.spender, args.amount);
    } else if (args.method == 'transfer_from') {
        return transfer_from(args.sender, args.recipient, args.amount);
    } else {
        throw 'method not found';
    }
}

function main() {
    const args = JSON.parse(PVM.load_args());
    PVM.debug(JSON.stringify(args));
    return _main(args) || '';
}

File: built-in-services/riscv/examples/dex/erc20.c
#include <stdint.h>

#include "./UsefulBuf.h"
#include "./cJSON.h"
#include "./pvm.h"
#include "./pvm_extend.h"
#include "./pvm_structs.h"

pvm_bytes_t _load_args() {
  uint8_t buf[2048];
  uint64_t args_size;

  pvm_load_args(buf, &args_size);
  return pvm_bytes_nbytes(buf, args_size);
}

pvm_bytes_t _balance_key(pvm_bytes_t *account) {
  pvm_bytes_t key = pvm_bytes_str("balance: ");
  pvm_bytes_append(&key, account);
  return key;
}

pvm_bytes_t _caller() {
  uint8_t buf[50];
  pvm_caller(buf);
  return pvm_bytes_nbytes(buf, 50);
}

pvm_u64_t total_supply() {
  pvm_bytes_t supply_key = pvm_bytes_str("supply");
  return pvm_u64_new(pvm_get_u64(&supply_key));
}

void set_balance(pvm_bytes_t *account, pvm_u64_t amount) {
  pvm_assert_not_null(account, "set balance account null");
  pvm_assert_not_empty(account, "set balance account empty");

  pvm_bytes_t key = _balance_key(account);
  pvm_set_u64(&key, pvm_u64_raw(amount));
}

void init(const char *name, const char *symbol, pvm_u64_t supply) {
  pvm_assert(!pvm_is_init(), "init can only be invoked by deploy function");

  pvm_bytes_t name_key = pvm_bytes_str("name");
  pvm_bytes_t symbol_key = pvm_bytes_str("symbol");
  pvm_bytes_t supply_key = pvm_bytes_str("supply");

  pvm_set_str(&name_key, name);
  pvm_set_str(&symbol_key, symbol);
  pvm_set_u64(&supply_key, pvm_u64_raw(supply));

  pvm_bytes_t caller = _caller();
  set_balance(&caller, supply);
}

pvm_u64_t balance_of(pvm_bytes_t *account) {
  pvm_assert_not_null(account, "balance of account null");
  pvm_assert_not_empty(account, "balance of account empty");

  pvm_bytes_t key = _balance_key(account);
  return pvm_u64_new(pvm_get_u64(&key));
}

void _transfer(pvm_bytes_t *sender, pvm_bytes_t *recipient, pvm_u64_t amount) {
  pvm_assert(pvm_u64_raw(amount) > 0, "transfer amount must be positive");

  pvm_u64_t from_balance = balance_of(sender);
  pvm_u64_t to_balance = balance_of(recipient);

  from_balance = pvm_u64_sub(from_balance, amount);
  to_balance = pvm_u64_add(to_balance, amount);

  pvm_set_u64(sender, pvm_u64_raw(from_balance));
  pvm_set_u64(recipient, pvm_u64_raw(to_balance));
}

void transfer(pvm_bytes_t recipient, pvm_u64_t amount) {
  pvm_bytes_t caller = _caller();
  _transfer(&caller, &recipient, amount);
}

pvm_bytes_t _approve_key(pvm_bytes_t *owner, pvm_bytes_t *spender) {
  pvm_bytes_t key = pvm_bytes_alloc(100);
  pvm_bytes_append(&key, owner);
  pvm_bytes_append(&key, spender);

  return key;
}

void _approve(pvm_bytes_t *owner, pvm_bytes_t *spender, pvm_u64_t amount) {
  pvm_bytes_t key = _approve_key(owner, spender);
  pvm_set_u64(&key, pvm_u64_raw(amount));
}

void approve(pvm_bytes_t *spender, pvm_u64_t amount) {
  pvm_bytes_t caller = _caller();
  _approve(&caller, spender, amount);
}

pvm_u64_t allowances(pvm_bytes_t *owner, pvm_bytes_t *spender) {
  pvm_bytes_t key = _approve_key(owner, spender);
  return pvm_u64_new(pvm_get_u64(&key));
}

void transfer_from(pvm_bytes_t *sender, pvm_bytes_t *recipient,
                   pvm_u64_t amount) {
  pvm_bytes_t caller = _caller();
  pvm_u64_t before_allowance = allowances(sender, recipient);
  pvm_u64_t after_allowance = pvm_u64_sub(before_allowance, amount);

  _transfer(sender, recipient, amount);
  _approve(sender, &caller, after_allowance);
}

pvm_bytes_t _cjson_get_str_bytes(cJSON *json, const char *item_name) {
  const cJSON *item = NULL;

  item = cJSON_GetObjectItemCaseSensitive(json, item_name);
  pvm_assert(cJSON_IsString(item), "item isn't string");
  pvm_assert(item->valuestring != NULL, "item is null");

  return pvm_bytes_str(item->valuestring);
}

const char *_cjson_get_str(cJSON *json, const char *item_name) {
  pvm_bytes_t item = _cjson_get_str_bytes(json, item_name);
  return pvm_bytes_get_str(&item);
}

pvm_bytes_t _cjson_get_str_bytes_or_empty(cJSON *json, const char *item_name) {
  const cJSON *item = NULL;

  item = cJSON_GetObjectItemCaseSensitive(json, item_name);
  pvm_assert(cJSON_IsString(item), "item isn't string");

  if (item->valuestring == NULL) {
    return pvm_bytes_empty();
  } else {
    return pvm_bytes_str(item->valuestring);
  }
}

pvm_u64_t _cjson_get_u64(cJSON *json, const char *item_name) {
  const cJSON *item = NULL;

  item = cJSON_GetObjectItemCaseSensitive(json, item_name);
  pvm_assert(cJSON_IsNumber(item), "item isn't number");

  uint64_t u64;
  memcpy(&u64, &item->valuedouble, sizeof(uint64_t));
  return pvm_u64_new(u64);
}

int main() {
  pvm_bytes_t json = _load_args();
  cJSON *args = cJSON_Parse(pvm_bytes_get_str(&json));
  pvm_assert(args != NULL, "invalid json args");

  const char *method = _cjson_get_str(args, "method");
  if (0 == strcmp(method, "init")) {
    const char *name = _cjson_get_str(args, "name");
    const char *symbol = _cjson_get_str(args, "symbol");
    pvm_u64_t supply = _cjson_get_u64(args, "supply");

    init(name, symbol, supply);
  } else if (0 == strcmp(method, "total_supply")) {
    pvm_ret_u64(pvm_u64_raw(total_supply()));
  } else if (0 == strcmp(method, "balance_of")) {
    pvm_bytes_t account = _cjson_get_str_bytes_or_empty(args, "account");

    if (!pvm_bytes_is_empty(&account)) {
      pvm_ret_u64(pvm_u64_raw(balance_of(&account)));
    } else {
      pvm_bytes_t caller = _caller();
      pvm_ret_u64(pvm_u64_raw(balance_of(&caller)));
    }
  } else if (0 == strcmp(method, "transfer")) {
    pvm_bytes_t recipient = _cjson_get_str_bytes(args, "recipient");
    pvm_u64_t amount = _cjson_get_u64(args, "amount");

    transfer(recipient, amount);
  } else if (0 == strcmp(method, "allowances")) {
    pvm_bytes_t owner = _cjson_get_str_bytes(args, "owner");
    pvm_bytes_t spender = _cjson_get_str_bytes(args, "spender");

    allowances(&owner, &spender);
  } else if (0 == strcmp(method, "approve")) {
    pvm_bytes_t spender = _cjson_get_str_bytes(args, "spender");
    pvm_u64_t amount = _cjson_get_u64(args, "amount");

    approve(&spender, amount);
  } else if (0 == strcmp(method, "transfer_from")) {
    pvm_bytes_t sender = _cjson_get_str_bytes(args, "sender");
    pvm_bytes_t recipient = _cjson_get_str_bytes(args, "recipient");
    pvm_u64_t amount = _cjson_get_u64(args, "amount");

    transfer_from(&sender, &recipient, amount);
  } else {
    // Alway false
    pvm_assert(0 > 1, "method not found");
  }

  return 0;
}
